src/core/src/pass/serialize.cpp

```diff
@@ -107,13 +129,67 @@ class ConstantWriter {
             return found->second.first;
         }
 
-        m_binary_output.write(ptr, size);
+        write_with_optional_fp16_compression(ptr, size, compress_to_fp16, src_type);
         m_hash_to_file_positions.insert({hash, {offset, static_cast<void const*>(ptr)}});
 
         return offset;
     }
 
 private:
+    void write_with_optional_fp16_compression(const char* ptr,
+                                              size_t size,
+                                              bool compress_to_fp16 = false,
+                                              ov::element::Type src_type = ov::element::dynamic) {
+        if (!compress_to_fp16) {
+            m_binary_output.write(ptr, size);
+        } else {
+            assert(size % 2 == 0);
+            size_t compressed_size;
+            // std::unique_ptr<char[]> fp16_buffer(new char[compressed_size]); // TODO: works?
+            // auto compressed_ptr = fp16_buffer.get();
+            auto fp16_buffer = compress_data_to_fp16(ptr, size, src_type, &compressed_size);
+            m_binary_output.write(fp16_buffer.get(), compressed_size);
+            // Compressed data is disposed
+        }
+    }
+
+    std::unique_ptr<char[]> compress_data_to_fp16(const char* ptr,
+                                                  size_t size,
+                                                  ov::element::Type src_type,
+                                                  size_t* compressed_size) {
+        if (src_type == ov::element::f32) {
+            *compressed_size = size / 2;
+            auto new_ptr = std::unique_ptr<char[]>(new char[size / 2]);
+            auto dst_data = reinterpret_cast<ov::float16*>(new_ptr.get());
+            auto src_data = reinterpret_cast<const float*>(ptr);
+            ngraph::runtime::reference::convert_from_f32_to_f16_with_clamp(src_data, dst_data, size / 4);
+            return new_ptr;
+        } else if (src_type == ov::element::f64) {
+            *compressed_size = size / 4;
+            auto new_ptr = std::unique_ptr<char[]>(new char[size / 4]);
+            auto dst_data = reinterpret_cast<ov::float16*>(new_ptr.get());
+            auto src_data = reinterpret_cast<const float*>(ptr);
+
+            // FIXME: This is SLOW stub implemenation!
+            //#pragma omp parallel for
```

Comment:
again - is it optimized or SLOW?

Reply:
We optimized only fp32 -> fp16 for both memory and performance, exept arm where we have optimized memory only as well as fp64 on all platforms.